(Solution for 2.2)  Enhance terminal querying with thread-safe buffering and improved async handling

TAJ-18 opened the pull request:
This PR introduces enhancements to the terminal querying mechanism, addressing challenges related to buffering, thread safety, and async handling. The changes ensure reliable terminal interactions in both synchronous and asynchronous environments.

Current Implementation

1)Reading Responses:
-When querying the terminal, a command is sent, and the response is read from the STDIN buffer.
-Non-response data in the STDIN buffer must be buffered separately to avoid mixing.

2)readansi Function:
-The terminal.lua library provides a readansi function (matching system.readansi) to handle reading and buffering.
-This ensures input handling is managed within the library, preventing conflicts with other parts of the application.

3)Thread Safety:
-Querying the terminal requires blocking behavior between sending the command and reading the response.
-The library allows passing a blocking sleep function to initialize to handle this case.



Proposed Improvements

1)Enhanced Buffering Mechanism:
-Implement a ring buffer or dedicated input queue to handle mixed data in the STDIN buffer.
-This ensures non-response data is preserved and processed correctly.

2)Thread-Safe Querying:
-Introduce a mutex or semaphore to ensure only one thread can query the terminal at a time.
-This prevents race conditions and ensures responses are not mixed.

3)Integration with Standard I/O:
-Add a flush_buffer function to flush buffered data back to STDIN.
-Provide a read_custom function to read from the buffered input, ensuring compatibility with standard I/O functions.

4)Blocking and Yielding Sleep:
-Improve the handling of blocking and yielding sleep functions.
-Allow the library to dynamically switch between blocking and yielding sleep based on the context (e.g., blocking for queries, yielding for general input).

Tieske commented on src/terminal/init.lua:
```diff
@@ -12,6 +12,122 @@
+  }
+end
+
+-- Initialize the buffer (Adjust size as needed)
+local input_buffer = create_ring_buffer(1024)
+
+-- Mutex implementation for thread safety
+local mutex = {
+  locked = false,
+  waiters = {}
+}
+
+function mutex.lock()
+  while mutex.locked do
+      table.insert(mutex.waiters, coroutine.running())
+      coroutine.yield()
+  end
+  mutex.locked = true
+end
+
+function mutex.unlock()
+  mutex.locked = false
+  local next_waiter = table.remove(mutex.waiters, 1)
+  if next_waiter then
+      coroutine.resume(next_waiter)
+  end
+end
```
I don't think this can work. Any mutex that has to yield to a scheduler, is tightly coupled to that scheduler. In other words; the scheduler must provide the mutex to use.

It's the exact same thing with the `sleep` functions. The scheduler provides them, and we pass them into the terminal module as parameters.

Have a look at [copas](https://github.com/lunarmodules/copas). It [the reference](https://lunarmodules.github.io/copas/reference.html) you'll find `copas.pause`, `mutex`, and `semaphore`. Or as another example; [OpenResty](https://github.com/openresty/lua-nginx-module), which provides `ngx.sleep`, and also some classes for locks and semaphores. So we cannot define our own, but must use what is provided.

TAJ-18 replied:
@Tieske,
You’re absolutely right—this custom mutex implementation is tightly coupled to the scheduler, and it’s not ideal to define our own in this context. I see now that relying on scheduler-provided primitives (like copas.mutex or ngx.semaphore) is the way to go, as it ensures compatibility and avoids reinventing the wheel.

I’ll refactor this to use the scheduler’s mutex or semaphore instead of the custom implementation. Thanks for pointing me to copas and OpenResty as references—I’ll take a closer look at those and adjust the code accordingly.

Tieske commented on src/terminal/init.lua:
```diff
@@ -12,6 +12,122 @@
+  if not input_buffer.is_empty() then
+      return input_buffer.pop()
+  end
+  return io.read()
```
`io.read` is off-limits, since it will block. All input must be handled through `system.readansi()`.

TAJ-18 replied:
@Tieske,
You’re absolutely right—io.read is blocking and shouldn’t be used here. I’ll refactor this to use system.readansi() for all input handling to ensure non-blocking behavior. Thanks for catching that!

Tieske commented on src/terminal/init.lua:
```diff
@@ -12,6 +12,122 @@
+end
+
+-- Thread-safe query function
+function terminal.query(cmd)
+  mutex.lock()
```
this is interesting, a threadsafe query method. Currently there is only 1 query implemented, the cursor position. It could be generalized. But that would have to be in the `terminal.input` module.

the parameters are not complete I think, since there needs to be a cmd to be sent, but also a pattern to validate to match on the input, to ensure we captured the response to the specific command.

TAJ-18 replied:
Hey @Tieske,
Good point—the terminal.query function should indeed be more generalized to handle any command and response pattern, not just the cursor position. I’ll move this logic to the terminal.input module and update the function to accept both a cmd and a response_pattern to validate the input. This will ensure we capture the correct response for the specific command.

Thanks for the feedback—I’ll make these changes and share an updated version soon!

Cheers,
TAJ-18

TAJ-18 commented on src/terminal/init.lua:
```diff
@@ -12,6 +12,122 @@
+
+
+-- Ring buffer implementation
+local function create_ring_buffer(size)
```
Hey @Tieske,
I chose the ring buffer design because it’s more efficient for handling continuous input streams, especially when working with fixed-size constraints. It avoids dynamic memory growth and ensures O(1) push/pop operations, which is great for performance.

The existing table approach works, but it can lead to inefficiencies when removing elements from the start (e.g., shifting elements in the table). The ring buffer simplifies buffer management by reusing space in a circular manner, which is especially useful for terminal input where we want to prioritize recent data and avoid unbounded memory usage.

If you think the existing table approach is better suited for this use case, I’m happy to adjust or discuss further!

Cheers,
TAJ-18

TAJ-18 commented:
Hello  @Tieske ,
 I have updated the code based on your feedback. Please review the changes:
 - Replaced single-table buffer with a ring buffer for better efficiency.
 - Removed custom mutex and used scheduler-provided mutex for thread safety.
- Ensured all input handling uses `system.readansi` to avoid blocking.
 - Generalized `terminal.query` to handle custom commands and response patterns.
 -  Thank you for your guidance!

Tieske commented on src/terminal/init.lua:
```diff
@@ -127,21 +119,19 @@ function terminal.sleep(is_blocking)
-  _DESCRIPTION = "Cross platform terminal library for Lua (Windows/Unix/Mac)",
+    _VERSION = "0.0.1",
+    _COPYRIGHT = "Copyright (c) 2024-2024 Thijs Schreijer",
+    _DESCRIPTION = "Cross platform terminal library for Lua (Windows/Unix/Mac)",
```
Can we undo the whitespace changes? it should be a 2 space indent (see file `.editorconfig`)

Tieske replied:
Also good practice is to keep commits clean and atomic. So make 1 change per commit. Don't mix multiple features, fixes, or anything else in a single commit.
I just added the `CONTRIBUTING.md` file today, might be worth while to read that.

Tieske commented on src/terminal/init.lua:
```diff
@@ -56,45 +54,40 @@ end
-  mutex.locked = true
-end
+-- Use scheduler-provided mutex (e.g., copas.mutex or ngx.semaphore)
+local mutex = require("copas.mutex") -- or ngx.semaphore
```
I'm not convinced about wrapping this in a mutex. The main problem is the scope of the input being read. A single key (eg. "press any key" type thing), or how about a readline function? then all input until the user presses enter should be within the mutex. Then again; then during the readline, it would become impossible to query the termnial for anything (querying cursor pos might come in handy during a readline).

So I don't know...

Tieske commented:
Have you tested this code? Does it even work? there are so many failures that it makes me doubt...
